### ISSUE TITLE
OsmChange for OSM API upload

### ISSUE THREAD

**Issue by mmd-osm** [osmcode/libosmium#239]
I'm currently looking into a speeding up the API 0.6 upload call via bulk operations (https://github.com/zerebubuth/openstreetmap-cgimap/issues/140), and are very happy with the support from libosmium for prototyping purposes.

However, there are several issues I'm struggling with a bit at the moment:

- `<delete>` element can have an additional attribute "if-unused", which is also not exposed via libosmium. The purpose of this attribute is to skip some operations which would otherwise trigger failures and would abort the whole upload call.

- osmium::Node, Way and Relation don't expose details,  if they have been part of a `<create>`, `<modify>` or `<delete>` section in an OsmChange file. Usually, one could handle this by referring to the version of an object. However, in the case of `<create>`, users can put in arbitrary values for the version number and the server is expected to still identify this as a create operation. In addition, a version number "1" doesn't mean to create a new object, but to apply a change to an object on the server, which currently has version 1 on the database. See [link](https://wiki.openstreetmap.org/wiki/API_v0.6#Diff_upload:_POST_.2Fapi.2F0.6.2Fchangeset.2F.23id.2Fupload) to the API 0.6 documentation.

- OsmChange document validation. Rails test suite contains a test case with invalid elements. This is not detected as invalid input by libosmium:

```xml
<osmChange version='0.6'>
  <ping>
   <node id='1' lon='1' lat='1' changeset='416' />
  </ping>
</osmChange>
```

- libosmium expects a version=0.6 Attribute on the header. In an upload scenario this is not mandatory and is actually ignored. Libosmium should not reject such input by raising an error.

Reference: 
* https://wiki.openstreetmap.org/wiki/OsmChange
* https://wiki.openstreetmap.org/wiki/API_v0.6#Diff_upload:_POST_.2Fapi.2F0.6.2Fchangeset.2F.23id.2Fupload

I wonder what the recommended approach would be to have more details about the original OsmChange XML document available in osmium::Node, Way and Relation objects.

TIA!

**Comment by mmd-osm** [comment]
I also looked at providing a copy of xml_input_format.hpp which processes OsmChange messages according to my needs. Then there's the issue that I cannot register multiple XML handlers via `ParserFactory::instance().register_parser`, and I also cannot extend osmium::io::fileformat.

I could in theory just avoid including xml_input.hpp. Unfortunately, that's not a good design, as it will make it possible in the future to use xml_input.hpp in unrelated parts of the code.

Then there's the issue with the  "if-unused" as mentioned before. I just cannot enhance existing Nodes/Ways/Relations with this flag. Still, I don't see any way to pass such a flag on a per-object basis from a Parser instance to my Handler implementation.

**Comment by joto** [comment]
1. Does the `if-unused` on the `delete` really have to be different on different objects? Or would the usual case be to have this on either all or none of the objects anyway? If that is the case, we could simply add an option for the XML file type.
2. create vs. modify: Maybe you can have a look at the server code and see whether the distinction create vs. modify has to actually be made here or if the server could deduce this from the context (ie whether it already has a version of the object with this version number). Maybe this is something we can solve on the server side?
3. document validation: The Osmium XML parser is a bit more lenient than it should really be. On the other hand being a bit forgiving here helps with forward compatibility and doesn't really hurt much. I have ti think about this.
4. Ignoring the version attribute doesn't strike me as a good idea, I would consider this a bug in the API. Sure we could work around this in Osmium, but it will bite us in the future, when the API moves to a new version.

**Comment by joto** [comment]
I only now realized that you are not trying to use the API, but to reimplement it on the server side. This makes things more complicated. I don't think either approach (extending the existing XML parser code to work also in this case, or adding a new XML parser specifically for this case) is the right one here. Both have the problem that you need to extend basic building blocks of Osmium, specifically the OSMObject class, to add some bits for a very specific use case. This creates a bottomless pit of hard-to-understand code and special cases. I am very reluctant to go in that direction.

If it is possible to change the API in a way that for (1) and (2) we can live with current semantics of the OSM Change file format, then we can just use the Osmium code. This would be the best option in my opinion. So first you'd have to figure out if the `if-unsed` attribute is actually used and if the create vs. modify is really needed. If in *actual API use* this doesn't matter, maybe we can get rid of it. Or maybe we can change this mechanism in some way. We are in the lucky position that we know about all API use and we can also change the API. And it wouldn't be the first time that an API contains something that was needed in theory but that in practice nobody ever used or something like that. Maybe we can simplify the API instead of complicating Osmium.

If this doesn't work then we have to go a different route. I could imagine adding another layer between the XML parser itself and the rest of Osmium. Separate the parts that are specifically about XML from the parts that create the OSMObjects. Then one part can be used in a different way by the "API parser". But I don't know enough about the API to see where this might be going.

**Comment by mmd-osm** [comment]
> I only now realized that you are not trying to use the API, but to reimplement it on the server side.

Yes, exactly, that's the idea. Current rails based implementation is quite slow, due to many single row operations. The redesign I'm working on uses mass database operations instead where possible and moves the whole changeset upload to cgimap. Local tests show a factor 100-200 speed up so far.

> Both have the problem that you need to extend basic building blocks of Osmium, specifically the OSMObject class, to add some bits for a very specific use cas

Right. I'm abusing some fields available now in OSMObject, but that's not really a good solution. I don't know if we could provide something like a `set_extra_attributes` methods instead, or more generally, an additional key-value-store similar to tags(), which could be used to store additional data per OSMObject, which is not part of the actual OSM data. Would this be an acceptable approach for you, which is not too much of a special case and keeps overall memory requirements stable?

Otherwise I agree to the following statement:
> extending the existing XML parser code to work also in this case, or adding a new XML parser specifically for this case)

My question would be, how an extension mechanism could look like. In my code example, I'm replacing the libosmium parser with my adjusted version. I would appreciate some official way to tell libosmium: hey that's my parser for filetype "xyz", please use the following Parser implementation in this case. I don't think my implementation itself would be suitable to be merged into the main libosmium code, it's way too special case.

> So first you'd have to figure out if the if-unsed attribute is actually used and if the create vs. modify is really needed. If in actual API use this doesn't matter, maybe we can get rid of it. 

Unfortunately, we can't. It used productively by iD editor at this time, and they specifically requested this parameter some years back.

> Or maybe we can change this mechanism in some way.

I'm a bit worried that we're creating too many dependencies here to move the Rails code to cgimap. If other apps like iD need to be synchronized to those changes, this whole thing will never go live. My idea is to have an ideally 100% compatible changeset upload, only much faster on C++. No need for any consumer to change anything.

> I could imagine adding another layer between the XML parser itself and the rest of Osmium. 

I think, for my use case, being able to provide (and switching) to an externally provided Parser implementation would already be a huge step forward. I don't really mess too much with the existing libosmium implementation and make it unmaintainable.

So summing up my wish list:

* Extra attributes (key/value) for OSMObject
* Flexible Parser selection, making it easy to use externally provided Parser implementation.

**Comment by joto** [comment]
Sorry, I can't see any way to add extra (generic) attributes to OSMObjects. This code is organized in a way that makes it highly efficient, any change I can think of would make it clunky and complicated.

Flexible parser selection hinges on changing the [file_format](https://github.com/osmcode/libosmium/blob/master/include/osmium/io/file_format.hpp#L42-L52) enum class. Problem is: An enum isn't all that flexible. Changing this from an enum to something else would potentially break user code, so I can't really change that without a new major version of libosmium. I could provide some types like `custom1`, `custom2` and then the application could use those. Horrible hack, but that would work. The [file format detection from the suffix](https://github.com/osmcode/libosmium/blob/master/include/osmium/io/file.hpp#L179) would not work for these format, but they can be set "manually" from the application.

> I'm a bit worried that we're creating too many dependencies here to move the Rails code to cgimap.

Understandable. But if you can't get this somewhat cleanly into libosmium the whole thing is questionable, too. I wouldn't exclude changes to the API if that could get us somewhere. You would have to do those changes in the Rails port first und only if everything works, put that into C++. I would argue that these (minimal) incompatible changes in the XML in the API are warts anyway and it makes sense to at least think about whether we can get rid of them.

**Comment by mmd-osm** [comment]
I somehow suspected that it would be extremely hard to get this working. It seems my use cases are often slightly outside the scope of libosmium, and unfortunately next to impossible to fit into the lib. Maybe that's the opportunity cost of having an extremly optimized lib that does tasks inside its scope pretty well (and much better than everone else).

 > file_format enum class

That's a bit of a pity. This enum class is used in several places/method calls, so a simple mapping to a string and using that instead would also not work without adjusting other method signatures as well. I agree that `custom1` and `custom2` are not that beautiful. Due to the lack of an extension concept for enum classes in c++, changing this would indeed be something for a next major release.

> But if you can't get this somewhat cleanly into libosmium the whole thing is questionable, too. 

I hope you're talking a bout the usage of libosmium for my use case, rather than giving up the idea of speeding up the whole changeset upload process. I agree, that I might have to look into running my own Expat XML parser implementation, substitue Osmium classes by a very lightweight clone for Nodes, Ways and Relations, and give up the buffering and compression handling. It's something I wanted to avoid in the first place due to the duplicate effort and reinventing the wheel.

>  I wouldn't exclude changes to the API if that could get us somewhere. [..] I would argue that these (minimal) incompatible changes in the XML in the API are warts anyway and it makes sense to at least think about whether we can get rid of them.

So this one is really challenging. So far we have:

* the version number in the header: JOSM, iD, Potlatch2 provide it, the wiki states it's not needed. Probbably not a big thing.
* `if-used` is not only used by iD, but also Potlatch 2 (I checked the sources again). I has real life semantic implications: a user won't get an conflict error message if another user already deleted an object in the meantime, and we're trying the same delete operation.
* operation (create/modify/delete): this one is also tricky, see discussion above.

Did I miss something?

> You would have to do those changes in the Rails port first and only if everything works, put that into C++

It really depends on the change. Removing `if-used` would e.g. immediately impact iD and Potlatch2 (with the effect of annoying users with more conflict resolution error messages). It's very unlikely that anyone would merge such a fix for Rails.

**Comment by joto** [comment]
This is getting off-topic here, but I am not saying you should remove `if-unused`. But I suspect that it is either set on all deletes or on none of them in which case it could be easily moved out of the XML file.

**Comment by mmd-osm** [comment]
I found the following snippet in P2 with alternating `if-used` elements: https://github.com/systemed/potlatch2/blame/ae1b2c6255eebb126de8cb95e4495243a32b859b/net/systemeD/halcyon/connection/XMLConnection.as#L253-L258. I have to admit that I find it a bit strange to try first without this flag (which would abort message processing anyway right away). @systemd possibly still remembers the reasoning behind this.

In iD, it's _always_ set (https://github.com/openstreetmap/iD/blob/015c8918f6ef1a6e2d8467a09d8b3b7359856729/modules/osm/changeset.js#L133). So here your assumption would be valid, we could move it outside of the XML document.

**Comment by mmd-osm** [comment]
The the time being, I'm closing this issue, as we more or less came to the conclusion that there's no easy short term "fix" available for libosmium, without breaking other apps - or - without having some changes to the main API.

That's not necessarily a permanent thing. Current dependency on libosmium is limited to certain parts of the coding, and switching in the future should still be possible, if the next major libosmium release covers some of the requriements. I'd be really happy to see a bit more flexibiility. It would be great if you could revisit this topic at some point in the future.

Thanks again for looking into this and providing valuable feedback.
